runcfanotify: replace pidfd_open/kill by reading /host/proc

alban opened the pull request:
pidfd_open(), os.FindProcess() and process.Signal() don't work for processes outside of the current pid namespace. Since we'd like IG to run in a container without hostPID=true, re-implement watchContainersTermination() with a periodic iteration on /proc (or /host/proc).

In preparation for #1658

alban commented on pkg/container-collection/container-collection.go:
```diff
@@ -422,11 +425,22 @@ func (cc *ContainerCollection) Close() {
-	// TODO: it's not clear if we want/can allow to re-initialize
-	// this instance yet, so we don't set cc.initialized = false.
-	cc.closed = true
+	// Similar to RemoveContainer() on all containers but without publishing
+	// events.
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(*Container)
+		for _, f := range c.cleanUpFuncs {
+			f()
+		}
+		cc.containers.Delete(c)
+		return true
+	})
```
The callback from SetFinalizer is called by the garbage collector and it might take many seconds before it happens. I think fd should be closed as soon as possible because:
1. In a scenario where containers are started and stopped quickly, ig might keep many file descriptor without needing it.
2. We have a test that checks for file descriptor leaks after the ContainerCollection is closed, and it fails if the mntns fd is not closed fast enough.
```
pkg/ig-manager/ig-manager_test.go:129:func TestClose(t *testing.T)
```

alban commented on pkg/container-collection/options.go:
```diff
@@ -698,6 +698,9 @@ func WithTracerCollection(tc TracerCollection) ContainerCollectionOption {
 		cc.cacheDelay = 2 * time.Second
 		cc.cachedContainers = &sync.Map{}
 
+		// container.cleanUpFuncs is called either at cc.RemoveContainer() or at cc.Close()
+		var cleanupMu sync.Mutex
```
The `cleanUpFuncs` for the same container can be called by two threads in parallel:
1. ContainerCollection's Close() calling WithTracerCollection's cleanUpFuncs calling the Container's cleanUpFuncs.
2. RemoveContainer's AfterFunc() having scheduled 2 seconds ago a call to the Container's cleanUpFuncs

The mutex solution is not very good because it requires cleanup functions to be idempotent and managing mutexes.

Another solution would be to keep the return value of AfterFunc() to be able to cancel the callback with Timer.Stop() but this means we have to store the timers somewhere. Should I do this? Or do you have a better code pattern to use here?

mauriciovasquezbernal replied:
I see. I put something together in https://github.com/inspektor-gadget/inspektor-gadget/commit/08d1afcc599144297deef0e814fce43f4b1eb9d3. 

The ideas are:
- Don't use `AfterFunc` but create a single goroutine that cleans up old cached containers. By doing so, we have a single place to acquire a mutex. 
- Remove cleanUpFuncs form container: Even if I like it because is generic, I think we don't need it and that generalization makes the code more difficult to write. 

I haven't tested it too much but I think it can work fine. What do you think?

alban replied:
I think that's a lot cleaner! I will pick your patch and test it.

alban commented on pkg/container-collection/container-collection.go:
```diff
@@ -426,6 +426,22 @@ func (cc *ContainerCollection) Close() {
+		unix.Close(c.mntNsFd)
+		c.mntNsFd = -1
+		return true
+	})
```
I refactored this based on Mauricio's patch below.

alban commented:
> Since this PR is not enough for removing `hostPID=true` (#1658), I will wait for general feedback on the approach described in [#1658 (comment)](https://github.com/inspektor-gadget/inspektor-gadget/issues/1658#issuecomment-1567114802) before addressing the review.

I have a way forward with https://github.com/inspektor-gadget/inspektor-gadget/issues/1658#issuecomment-1576468084 so I am picking up this PR again.

mauriciovasquezbernal commented on pkg/runcfanotify/runcfanotify.go:
```diff
@@ -244,131 +209,47 @@ func (n *RuncNotifier) AddWatchContainerTermination(containerID string, containe
 			return
-		}
+		case <-ticker.C:
 
```
nit: remove this empty line.

alban commented on pkg/runcfanotify/runcfanotify.go:
```diff
@@ -244,131 +209,47 @@ func (n *RuncNotifier) AddWatchContainerTermination(containerID string, containe
 			}
 
-			if err != nil {
+			for _, c := range n.containers {
+				if pids[c.pid] {
+					// container still running
+					continue
+				}
+
```
Right, we don't handle this case. I think it is low risk.

alban commented:
Thanks for the review! I updated the PR and I will merge it once the CI passes.
